src/overlay/Floodgate.h

```diff
@@ -44,11 +44,30 @@ class Floodgate
                     Peer::pointer peer);
     };
 
+    bool mWaitingDemand{false};
     std::map<Hash, FloodRecord::pointer> mFloodMap;
     Application& mApp;
     medida::Counter& mFloodMapSize;
     medida::Meter& mSendFromBroadcast;
+    medida::Meter& mMessagesAdvertized;
+    medida::Meter& mMessagesFulfilled;
     bool mShuttingDown;
+    std::map<Hash, std::pair<int, VirtualClock::time_point>> mDemandHistory;
+    bool shouldFloodLazily(StellarMessage const& msg, Peer::pointer peer) const;
+    VirtualTimer mDemandTimer;
+
+    static constexpr std::chrono::milliseconds FLOOD_DEMAND_BACKOFF_RATE{200};
+
+    // This is used for linear backoff.
+    // If we get to the (MAX_RETRY_COUNT + 1)-th demand,
+    // it has been (1 + 2 + ... + MAX_RETRY_COUNT + (MAX_RETRY_COUNT + 1)) *
+    // FLOOD_DEMAND_BACKOFF_RATE since the first demand, which is
+    // (15 + 2) * (15 + 1) / 2 * 500 ms = 68 seconds.
+    // It's likely that no one has the txn, so we might as well give up.
+    int const MAX_RETRY_COUNT = 15;
+
+    static constexpr std::chrono::milliseconds
+        MAX_DEMAND_HISTORY_RETENTION_DURATION{100000};
```

Comment:
Is this needed? Can't you remove the demand after you hit MAX_RETRY_COUNT, since we don't need to track the demand once we hit the limit?

Reply:
I went with this approach since I thought about the case when fewer than `MAX_RETRY_COUNT` peers send us a transaction hash & none of them responds. For instance, if a malicious node sends us a hash for a transaction that doesn't exist and if that's the only node to send us that hash, then we'll never hit `MAX_RETRY_COUNT`.

Reply:
I changed it to ` MAX_DELAY_DEMAND * MAX_RETRY_COUNT * 2`, which is `2 sec * 15 * 2 = 1 min`.